Add timezone support to time intervals.

benridley opened the pull request:
This PR addresses #2771 by adding a `time_zone` field to time interval definitions, which accepts an IANA time zone name (like 'Australia/Sydney') and if supplied means that the time intervals are defined inside that time zone.

On Windows machines the only time zones that are accepted are 'Local' and 'UTC', but I think that is probably enough for 99% of users who just want to use their local server time zone. I have noted this in the documentation, and also added some more info to the error message when a timezone fails to load on Windows mentioning that the user may need to supply their own time zone database.

I think it's also probably worth changing the default to be the server's local time rather than UTC as this would probably be unexpected for most users, but it is breaking behavior so I'm not sure how we should proceed with that but open to the ideas.

Feedback on the approach (particularly dealing with Windows) is more than welcome. Cheers!

benridley commented:
It seems I've fixed the tests that were failing, but I'm not sure what's going wrong with the 'mixin' part of the CI pipeline, it looks unrelated to my changes...

PhilipGough commented:
@benridley looks like this was addressed in #2785 so a rebase on top of main should resolve this for you

benridley commented:
Thanks @PhilipGough, I've just rebased so we'll see how it goes.

sylr commented on timeinterval/timeinterval.go:
```diff
@@ -405,6 +457,13 @@ func clamp(n, min, max int) int {
 
 // ContainsTime returns true if the TimeInterval contains the given time, otherwise returns false.
 func (tp TimeInterval) ContainsTime(t time.Time) bool {
+	var location *time.Location
+	if tp.TimeZone == nil {
+		location = time.UTC
+	} else {
+		location = tp.TimeZone.Location
+	}
+	t = t.In(location)
```
```suggestion
	if tp.TimeZone != nil {
		t = t.In(tp.TimeZone.Location)
	}
```

benridley replied:
Nice catch, thanks

sylr commented on timeinterval/timeinterval_test.go:
```diff
@@ -434,7 +480,7 @@ func TestYamlUnmarshal(t *testing.T) {
 		}
 		for _, ts := range tc.contains {
-			_t, _ := time.Parse(time.RFC822, ts)
+			_t, _ := time.Parse(time.RFC822Z, ts)
```
I wouldn't have changed that. The purpose of those tests is to ensure non regression so if we change all the test suit we can't be sure we didn't break things. Also I don't see the benefit to use numeric timezones instead of the original short codes. I would revert this and use AEDT / AEST for the new test cases related to timezones.

benridley replied:
I changed these because the original format (RFC822) doesn't actually work the way I thought it did (I wrote the original tests as well). It doesn't actually support timezones like "AEDT". I realized this because it was failing on my new tests, until I read the RFC and as it turns out it only supports 4 static timezones. So I changed to this format so I could test with more timezones, including changing between daylight savings etc.

benridley replied:
To add some clarification to this (in case it's a blocker for merging), using RFC822 with timezones like AEDT / AEST just falls back to UTC, because it doesn't understand those time zones at all. The new format (RFC822Z) allows us to test the behaviour correctly by using numeric offsets.

simonpasquier commented on timeinterval/timeinterval.go:
```diff
@@ -165,6 +172,31 @@ var monthsInv = map[int]string{
+
+	loc, err := time.LoadLocation(str)
+	if err != nil {
+		if runtime.GOOS == "windows" && str != "Local" && str != "UTC" {
+			return fmt.Errorf("unable to load time location. Timezones other than 'Local' and 'UTC' may not be supported on Windows without using a custom timezone file: %w", err)
+		}
+		return err
```
I'm wondering if we could make the error message more targeted.

```suggestion
		if runtime.GOOS == "windows" {
			if zoneinfo := os.GetEnv("ZONEINFO"); zoneinfo != "" {
				return fmt.Errorf("%w (ZONEINFO=%q)", err, zoneinfo)
			}
			return fmt.Errorf("%w (on Windows platforms, you may have to pass the time zone database using the ZONEINFO environment variable, see https://pkg.go.dev/time#LoadLocation for details)", err)
		}
		return err
```

benridley replied:
Thanks @simonpasquier. 

Yeah I think that's a good idea. I've also been thinking more, and I'm wondering if we just go the extra distance and embed the timezone database on Windows platforms? Obviously it's not ideal because we can have stale timezone information between releases of AM, but I think that's probably better than nothing. What do you think?

simonpasquier replied:
I would defer it for a later PR if people feel that it's really needed. As you said, we're at risk of providing stale information.

beorn7 commented:
As discussed elsewhere, while "timezone" in the context of the TZ database indeed refers to a location (e.g. "Australia/Sydney"), the general understanding of "timezone" is more ambiguous. In many other contexts, a timezone is something like "PDT" or "PST" (or even "PST/PDT" :scream: ). As a concrete example particularly relevant for Go developers, look at the terminology in https://pkg.go.dev/time#Time .

Therefore, I suggest to use a word like "location". Even if you come from a TZ database context, you will still understand what that is, while those from other contexts will have a much easier life (not trying to enter "PDT" and wondering why it doesn't work).

benridley commented:
Sure @beorn7, I think that's a good suggestion and keeps the terminology consistent with Go's stdlib. Happy to add this.

simonpasquier commented:
Great feedback from @beorn7 (as usual) on location vs. time zone, +1 for me.

piotr1212 commented:
Hey, wondering if this is still being worked on as we'd need this feature. Also assuming this works well with daylight saving?

benridley commented:
Hey @piotr1212,

Yeah it's coming - I was mainly waiting on #2779 because I knew I'd need to rebase and fix this code up once that was merged. I'll probably work on it this weekend and get it tidied up.

And yes, it will work fine with daylight savings - Provided you aren't running AM on Windows.

benridley commented:
I've just rebased onto main, and also added in the following:
- Renamed Timezone/time_zone to Location/location as its a more accurate description and consistent with Go's stdlib (thanks @beorn7)
- Added in @simonpasquier's suggestion to be more specific with the Windows error and check.
- Fixed up the docs to use the 'location' keyword, and also added some more details about how it works.

sylr commented:
@simonpasquier could we go ahead and merge this please ?

sylr commented on notify/notify_test.go:
```diff
@@ -724,16 +724,20 @@ func TestMuteStageWithSilences(t *testing.T) {
 	muteIn := `
 ---
 - weekdays: ['monday:friday']
+  time_zone: 'Australia/Sydney'
```
```suggestion
  location: 'Australia/Sydney'
```

sylr commented on notify/notify_test.go:
```diff
@@ -724,16 +724,20 @@ func TestMuteStageWithSilences(t *testing.T) {
-- weekdays: ['saturday', 'sunday']`
+- weekdays: ['saturday', 'sunday']
+  months: ['November']
+  time_zone: 'Australia/Sydney'`
```
```suggestion
  location: 'Australia/Sydney'`
```

sylr commented on timeinterval/timeinterval.go:
```diff
@@ -33,6 +35,7 @@ type TimeInterval struct {
 	DaysOfMonth []DayOfMonthRange `yaml:"days_of_month,flow,omitempty" json:"days_of_month,omitempty"`
 	Months      []MonthRange      `yaml:"months,flow,omitempty" json:"months,omitempty"`
 	Years       []YearRange       `yaml:"years,flow,omitempty" json:"years,omitempty"`
+	Location    *Location         `yaml:"time_zone,flow,omitempty" json:"time_zone,omitempty"`
```
```suggestion
	Location    *Location         `yaml:"location,flow,omitempty" json:"location,omitempty"`
```

sylr commented on timeinterval/timeinterval_test.go:
```diff
@@ -378,6 +409,21 @@ var yamlUnmarshalTestCases = []struct {
+		in: `
+---
+- years: ['2020:2022']
+  time_zone: 'Australia/Sydney'
```
```suggestion
  location: 'Australia/Sydney'
```

benridley commented:
Thanks for those test corrections @sylr, I've just merged them.

simonpasquier reviewed (approved):
lgtm.

We could maybe add a blank import to https://pkg.go.dev/time/tzdata in cmd/alertmanager/main.go to embed a copy of the timezone database but it can be done later if we see fit.
